examples/python/pytket-qujax_qaoa.py

```diff
@@ -0,0 +1,130 @@
+# where $\alpha_{ij}$ are the coefficients.
+# Let's build our problem Hamiltonian with random coefficients and a set of pairs for a given number of qubits:
+
+n_qubits = 4
+hamiltonian_qubit_inds = [(0, 1), (1, 2), (0, 2), (1, 3)]
+hamiltonian_gates = [['Z', 'Z']] * (len(hamiltonian_qubit_inds))
+
+
+# Notice that in order to use the random package from jax we first need to define a seeded key
+seed = 13
+key = random.PRNGKey(seed)
+coefficients = random.uniform(key, shape=(len(hamiltonian_qubit_inds), ))
+
+print('Gates:\t', hamiltonian_gates)
+print('Qubits:\t', hamiltonian_qubit_inds)
+print('Coefficients:\t', coefficients)
+
+# ## Variational Circuit
+# Before constructing the circuit, we still need to select the mixing Hamiltonian. In our case, we will be using $X$ gates in each qubit, so $H_B = \sum_{i=1}^{n}X_i$, where $n$ is the number of qubits. Notice that the unitary $U(\beta)$, given this mixing Hamiltonian, is an $X$ rotation in each qubit with angle $\beta$.
+# As for the unitary corresponding to the problem Hamiltonian, $U(\gamma)$, it has the following form:
+# $$U(\gamma)=\prod_{(i, j) \in E}e^{-i\gamma\alpha_{ij}Z_iZ_j}$$ 
+# The operation $e^{-i\gamma\alpha_{ij}Z_iZ_j}$ can be performed using two CNOT gates with qubit $i$ as control and qubit $j$ as target and a $Z$ rotation in qubit $j$ in between them, with angle $\gamma\alpha_{ij}$.
+# Finally, the initial state used, in general, with the QAOA is an equal superposition of all the basis states. This can be achieved adding a first layer of Hadamard gates in each qubit at the beginning of the circuit.
+
+# With all the building blocks, let's construct the symbolic circuit using tket. Notice that in order to define the parameters, we use the ```Symbol``` object from the `sympy` package. More info can be found in this [documentation](https://cqcl.github.io/pytket/manual/manual_circuit.html#symbolic-circuits). In order to later convert the circuit to qujax, we need to return the list of symbolic parameters as well.
+
+def qaoa_circuit(n_qubits, depth):
+    circuit = Circuit(n_qubits)
+    p_keys = []
+    
+    # Initial State
+    for i in range(n_qubits):
+        circuit.H(i)
+        
+    for d in range(depth):
+        # Mixing Unitary
+        beta_d = Symbol(f"β_{d}")
+        for i in range(n_qubits):
+            circuit.Rx(beta_d, i)
+        p_keys.append(beta_d)
+        # Hamiltonian
+        gamma_d = Symbol(f"γ_{d}")
+        for index in range(len(hamiltonian_qubit_inds)):
+            pair = hamiltonian_qubit_inds[index]
+            coef = coefficients[index]
+            
+            circuit.CX(pair[0], pair[1])
+            circuit.Rz(gamma_d * coef, pair[1])
+            circuit.CX(pair[0], pair[1])
+            circuit.add_barrier(range(0, n_qubits))
+        p_keys.append(gamma_d)
+        
+    return circuit, p_keys
+
+depth = 3
+circuit, keys = qaoa_circuit(n_qubits, depth)
+
+keys
+
+# Let's check the circuit:
+
+render_circuit_jupyter(circuit)
```

Comment:
The numeric parameters are qubit-specific and the symbolic gamma parameters are layer-specific, so the numeric parameters are a feature of the ansatz inherited from the problem Hamiltonian.

It looks like Farhi et al also have the cost/gamma layer first, do you have a preference for having the mixing/beta layer first? @gabrielcqc

Reply:
Yes exactly, the fact that the gamma parameters are multiplied by some numerical factors is because of the coefficients of the problem Hamiltonian. In many demos that I've seen they usually set all the coefficients to 1 so you only have the gamma parameter, but I thought it would be better to do it more general.

As for the order of the unitaries, it is true that by definition the first is the gamma and then the beta, but in general these two unitaries commute so it should not be a problem. I have no preference for having the beta layer first, maybe it is a good idea to stick to the original notation (first gamma, then beta) to make it less confusing.

Reply:
Thanks both for the info on the parameter values. I've only seen QAOA in a maxcut context before.

As for the ordering I'll defer to @gabrielcqc on what is best. Is it true that the unitaries commute in general?

Reply:
Sorry I did not explain myself very clearly. In general they do not commute, if you consider any two hamiltonians it is most likely that they will not commute. However, people tend to select commuting Hamiltonians (at least in my experience). We have to take into account that to implement a QAOA we need to implement the unitary U^H, which can be very difficult in general, so that's why usually people use "simple" Hamiltonians.